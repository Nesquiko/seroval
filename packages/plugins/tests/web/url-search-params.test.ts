import { describe, it, expect } from 'vitest';
import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import URLSearchParamsPlugin from '../../web/url-search-params';

const EXAMPLE = new URLSearchParams('hello=world&foo=bar');

describe('URLSearchParams', () => {
  describe('serialize', () => {
    it('supports URLSearchParams', () => {
      const result = serialize(EXAMPLE, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('serializeAsync', () => {
    it('supports URLSearchParams', async () => {
      const result = await serializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof EXAMPLE>>(result);
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toJSON', () => {
    it('supports URLSearchParams', () => {
      const result = toJSON(EXAMPLE, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toJSONAsync', () => {
    it('supports URLSearchParams', async () => {
      const result = await toJSONAsync(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<typeof EXAMPLE>>(result, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('crossSerialize', () => {
    it('supports URLSearchParams', () => {
      const result = crossSerialize(EXAMPLE, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports URLSearchParams', () => {
        const result = crossSerialize(EXAMPLE, {
          plugins: [URLSearchParamsPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports URLSearchParams', async () => {
      const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports URLSearchParams', async () => {
        const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
          plugins: [URLSearchParamsPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports URLSearchParams', async () => new Promise<void>((resolve, reject) => {
      crossSerializeStream(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          resolve();
        },
        onError(error) {
          reject(error);
        },
      });
    }));
    describe('scoped', () => {
      it('supports URLSearchParams', async () => new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(EXAMPLE), {
          plugins: [URLSearchParamsPlugin],
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    });
  });
  describe('toJSON', () => {
    it('supports URLSearchParams', () => {
      const result = toCrossJSON(EXAMPLE, {
        plugins: [URLSearchParamsPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [URLSearchParamsPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toJSONAsync', () => {
    it('supports URLSearchParams', async () => {
      const result = await toCrossJSONAsync(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<Promise<typeof EXAMPLE>>(result, {
        plugins: [URLSearchParamsPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(URLSearchParams);
      expect(String(back)).toBe(String(EXAMPLE));
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports URLSearchParams', async () => new Promise<void>((resolve, reject) => {
      toCrossJSONStream(Promise.resolve(EXAMPLE), {
        plugins: [URLSearchParamsPlugin],
        onParse(data) {
          expect(JSON.stringify(data)).toMatchSnapshot();
        },
        onDone() {
          resolve();
        },
        onError(error) {
          reject(error);
        },
      });
    }));
  });
});
