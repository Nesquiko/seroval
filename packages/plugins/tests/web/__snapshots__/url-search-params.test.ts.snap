// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`URLSearchParams > crossSerialize > scoped > supports URLSearchParams 1`] = `"($R=>$R[0]=new URLSearchParams("hello=world&foo=bar"))($R["example"])"`;

exports[`URLSearchParams > crossSerialize > supports URLSearchParams 1`] = `"$R[0]=new URLSearchParams("hello=world&foo=bar")"`;

exports[`URLSearchParams > crossSerializeAsync > scoped > supports URLSearchParams 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=new URLSearchParams("hello=world&foo=bar")))($R["example"])"`;

exports[`URLSearchParams > crossSerializeAsync > supports URLSearchParams 1`] = `"$R[0]=Promise.resolve($R[1]=new URLSearchParams("hello=world&foo=bar"))"`;

exports[`URLSearchParams > crossSerializeStream > scoped > supports URLSearchParams 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`URLSearchParams > crossSerializeStream > scoped > supports URLSearchParams 2`] = `"($R=>($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new URLSearchParams("hello=world&foo=bar")))($R["example"])"`;

exports[`URLSearchParams > crossSerializeStream > supports URLSearchParams 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`URLSearchParams > crossSerializeStream > supports URLSearchParams 2`] = `"($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new URLSearchParams("hello=world&foo=bar"))"`;

exports[`URLSearchParams > serialize > supports URLSearchParams 1`] = `"new URLSearchParams("hello=world&foo=bar")"`;

exports[`URLSearchParams > serializeAsync > supports URLSearchParams 1`] = `"Promise.resolve(new URLSearchParams("hello=world&foo=bar"))"`;

exports[`URLSearchParams > toCrossJSONStream > supports URLSearchParams 1`] = `"{"t":22,"i":0,"f":{"t":26,"i":1,"s":1}}"`;

exports[`URLSearchParams > toCrossJSONStream > supports URLSearchParams 2`] = `"{"t":23,"i":0,"a":[{"t":26,"i":3,"s":2},{"t":25,"i":2,"s":{"t":1,"s":"hello=world&foo=bar"},"c":"seroval-plugins/web/URLSearchParams"}]}"`;

exports[`URLSearchParams > toJSON > supports URLSearchParams 1`] = `"{"t":{"t":25,"i":0,"s":{"t":1,"s":"hello=world&foo=bar"},"c":"seroval-plugins/web/URLSearchParams"},"f":31,"m":[]}"`;

exports[`URLSearchParams > toJSON > supports URLSearchParams 2`] = `"{"t":25,"i":0,"s":{"t":1,"s":"hello=world&foo=bar"},"c":"seroval-plugins/web/URLSearchParams"}"`;

exports[`URLSearchParams > toJSONAsync > supports URLSearchParams 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"t":1,"s":"hello=world&foo=bar"},"c":"seroval-plugins/web/URLSearchParams"}},"f":31,"m":[]}"`;

exports[`URLSearchParams > toJSONAsync > supports URLSearchParams 2`] = `"{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"t":1,"s":"hello=world&foo=bar"},"c":"seroval-plugins/web/URLSearchParams"}}"`;
